OpenRA.Game/SelectableExts.cs

```diff
@@ -43,14 +43,30 @@ public static int SelectionPriority(this Actor a)
 			}
 		}
 
-		public static Actor WithHighestSelectionPriority(this IEnumerable<Actor> actors)
+		public static Actor WithHighestSelectionPriority(this IEnumerable<Actor> actors, int2 selectionPixel)
 		{
-			return actors.MaxByOrDefault(a => a.Info.SelectionPriority());
+			return actors.MaxByOrDefault(a => CalculateActorSelectionPriority(a, selectionPixel));
 		}
 
-		public static FrozenActor WithHighestSelectionPriority(this IEnumerable<FrozenActor> actors)
+		public static FrozenActor WithHighestSelectionPriority(this IEnumerable<FrozenActor> actors, int2 selectionPixel)
 		{
-			return actors.MaxByOrDefault(a => a.Info.SelectionPriority());
+			return actors.MaxByOrDefault(a => CalculateActorSelectionPriority(a, selectionPixel));
+		}
+
+		static long CalculateActorSelectionPriority(Actor actor, int2 selectionPixel)
+		{
+			var centerPixel = new int2(actor.Bounds.X, actor.Bounds.Y);
+			var pixelDistance = (centerPixel - selectionPixel).Length;
+
+			return ((long)-pixelDistance << 32) + actor.Info.SelectionPriority();
+		}
+
+		static long CalculateActorSelectionPriority(FrozenActor actor, int2 selectionPixel)
+		{
+			var centerPixel = new int2(actor.Bounds.X, actor.Bounds.Y);
+			var pixelDistance = (centerPixel - selectionPixel).Length;
+
+			return ((long)-pixelDistance << 32) + actor.Info.SelectionPriority();
```

Comment:
Could you factor out a common method for the priority distance bits and pass in the specifics from the Actor/FrozenActor as parameters instead please?